File: 2/C/add_two_numbers_tb.c
#include "add_two_numbers.c"

int main() {
    struct ListNode *list1, *list2, *result;
    
    // // test 0: list constructor
    // struct ListNode* constructor_test = createList(10);
    // printList(constructor_test);
    // printf("%d\n", listToInt(constructor_test));
    // destroyList(constructor_test);
    
    // test 1 -- 342 + 456 = 807: [2 4 3] [5 6 4]  -->  [7 0 8]
    list1 = createList(342);
    list2 = createList(465);
    result = addTwoNumbers(list1, list2);
    printf("\nTEST 1\nExpected result: 7 0 8\nActual result:   ");
    printList(result);      
    destroyList(list1);
    destroyList(list2);
    destroyList(result);
    
    
    // test 2 -- 0 + 0 = 0: [0] [0]  -->  [0]
    list1 = createList(0);
    result = addTwoNumbers(list1, list1);
    printf("\nTEST 2\nExpected result: 0\nActual result:   ");
    printList(result);      
    destroyList(list1);
    destroyList(result);
    
    
    // test 3 -- 9 999 999 + 9 999 = 1 0 0 0 9 9 9 8: [9 9 9 9 9 9 9] [9 9 9 9]  -->  [8 9 9 9 0 0 0 1]
    list1 = createList(9999999);
    list2 = createList(9999);
    result = addTwoNumbers(list1, list2);
    printf("\nTEST 3\nExpected result: 8 9 9 9 0 0 0 1\nActual result:   ");
    printList(result);      
    destroyList(list1);
    destroyList(list2);
    destroyList(result);
 
}

File: 2/python/addTwoNumbers_tb.py
from addTwoNumbers import Solution
from listNode import ListNode

def main():
    # Need a mediator variable to access addTwoNumbers
    soln = Solution()

    # Test 1: 0 + 0 = 0
    list1 = ListNode(0)
    list2 = ListNode(0)
    result = soln.addTwoNumbers(list1, list2)
    assert result.toList() == [0]
        
    # Test 2: 1 + 2 = 3
    list1 = ListNode(1)
    list2 = ListNode(2)
    result = soln.addTwoNumbers(list1, list2)
    assert result.toList() == [3]
        
    # Test 3: 1 + 9 = 10
    list1 = ListNode(1)
    list2 = ListNode(9)
    result = soln.addTwoNumbers(list1, list2)
    assert result.toList() == [0, 1]
        
    # Test 4: 81 + 4218 = 4299
    list1 = ListNode(1)
    list1.next = ListNode(8)
    list2 = ListNode(8)
    list2.next = ListNode(1)
    list2.next.next = ListNode(2)
    list2.next.next.next = ListNode(4)
    result = soln.addTwoNumbers(list1, list2)
    assert result.toList() == [9, 9, 2, 4]
    
    print("PASSED ALL TESTS")
        
        
        
if __name__ == '__main__':
    main()